### PR TITLE
Improve dart-lang/markdown#51.

### PR DESCRIPTION
**Pull request by mehaase**
Improves dart-lang/markdown#51.

First, put a try/catch around attempting to parse Markdown.
Failure to parse should count as a failed test case, but
should not abort testing altogether.

Second, implement a primitive HTML comparison function that
is not sensitive to attribute order, meaningless whitespace,
etc.

The second change alone boosts CommonMark score from 232/599
to 383/599.

### CODE REVIEW

File: tool/common_mark_stats.dart [hunk 1]
```diff
@@ -62,6 +72,60 @@ void main() {
       '– ${pct}%');
 }
 
+/// Compare two DOM trees for equality.
+bool compareHtml(List<Element> expectedEls, List<Element> actualEls) {
```
Comment (kevmoo) [inline tool/common_mark_stats.dart]: don't use abbreviations here –> `expectedElements`, etc

File: tool/common_mark_stats.dart [hunk 2]
```diff
@@ -62,6 +72,60 @@ void main() {
...
+  if (expectedEls.length != actualEls.length) {
+    return false;
+  }
+
+  for (int childNum in new List.generate(expectedEls.length, (i) => i)) {
```
Comment (kevmoo) [inline tool/common_mark_stats.dart]: don't needlessly type `var childNum` etc...

Also ...just use `for (var childNum = 0; childNum < expectedElement.length; childNum++)`

File: tool/common_mark_stats.dart [hunk 3]
```diff
@@ -62,6 +72,60 @@ void main() {
...
+    if (expected is Element) {
+      if (expected.localName != actual.localName) {
+        return false;
+      }
+
+      if (expected.attributes.length != actual.attributes.length) {
+        return false;
+      }
+
+      List expectedAttrKeys = new List.from(expected.attributes.keys);
```
Comment (kevmoo) [inline tool/common_mark_stats.dart]: `var`

Also: `expected.attributes.keys.toList()`

File: tool/common_mark_stats.dart [hunk 4]
```diff
@@ -62,6 +72,60 @@ void main() {
...
+      expectedAttrKeys.sort();
+
+      List actualAttrKeys = new List.from(actual.attributes.keys);
+      actualAttrKeys.sort();
+
+      for (int attrNum in new List.generate(actualAttrKeys.length, (i) => i)) {
```
Comment (kevmoo) [inline tool/common_mark_stats.dart]: same as above

File: tool/common_mark_stats.dart [hunk 5]
```diff
@@ -62,6 +72,60 @@ void main() {
...
+        }
+
+        if (expected.attributes[expectedAttrKey] !=
+            actual.attributes.keys[actualAttrKey]) {
+          return false;
+        }
+      }
+    }
+
+    bool childrenEqual = compareHtml(expected.children, actual.children);
```
Comment (kevmoo) [inline tool/common_mark_stats.dart]: var

**Comment by kevmoo** [comment]
@mehaase do you have an example where a common-mark test is failing just because of the ordering of attributes?

**Comment by mehaase** [comment]
Just pushed an update that address your feedback.

**Comment by mehaase** [comment]
> @mehaase do you have an example where a common-mark test is failing just because of the ordering of attributes?

*I answered this question in a separate thread, but I realize that it's hard to keep track of the different threads, so I'm reposting that answer here.*

Test case 521 is failing because of the order of attributes. Here are the expected and actual outputs for that test case as it stands on `master` right now.

Expected:

    <p>foo</p>

Actual:

    <p>foo</p>

There are other test cases that rely on attribute order. Most of them are anchors that happen to have attributes in alphabetical order (`href` before `title`). The image test cases have `src` before `alt`, which will definitely cause test failures, but nobody noticed before because the image test cases were totally broken for other reasons.

The far, far more common issue, however, is insignificant whitespace. For example, test case 4:

Expected:

    <ul>
    <li>
    <p>foo</p>
    <p>bar</p>
    </li>
    </ul>

Actual:

    <ul><li>
    <p>foo</p>
    <p>bar</p></li></ul>

It's the exact same DOM. The whitespace between `<ul>` and `<li>`, for example, is insignificant. 

**Over 150 CommonMark tests fail due to insignificant differences in the markup.** Moreover, you'll *never get these tests right* unless you normalize the HTML output somehow. 

For example, jgm's repo (which is where you got the test cases from, right?) performs [a sophisticated normalization process](https://github.com/jgm/CommonMark/blob/master/test/normalize.py) so that HTML can be compared for significant differences.

**Comment by srawlins** [comment]
Hi @mehaase, sorry for the crazy long wait, can you fix merge conflicts?

**Comment by srawlins** [comment]
This is a great change. I can confirm that this patch improves our compliance from **237 of 613 -- 38.7% total** to **400 of 613 -- 65.3% total**.

@mehaase, if you can fix the conflicts surrounding the change from `validCount` to `nestedMap` (should be pretty simple), I can merge this.

**Comment by mehaase** [comment]
Thanks @srawlins. I fixed the merge conflicts.

**Comment by srawlins** [comment]
All tests pass, CommonMark coverage dramatically improved. Thanks!

**Comment by mehaase** [comment]
Thank you!

**Pull request merged**